### PR TITLE
Guidance document fix #9

### LINKED ISSUE #9: Design Guidelines
I'm not certain if this belongs in "Common Errors", or more something else. But I'd really like to expand on the trade-offs of fine- vs coarse-grained isolation.

For example:

```swift
class SomeClass {
    @MainActor
    var isolatedValue = 1
}
```

vs

```swift
actor SomeClass {
    @MainActor
    var isolatedValue = 1
}
```

vs

```swift
@MainActor
class SomeClass {
    var isolatedValue = 1
}
```

Especially to tie in the use of `nonisolated` for selectively opting-out of whatever domain your type is part of.

### PR DESCRIPTION
**Pull request by mattmassicotte**
I've begun work on a guidance document. Currently, this mostly lines up with the content from the "Migrate your app to Swift 6" WWDC video. However, I did get into the leaf vs root module problem.

I also have a lot more to do here, but I thought it would be useful to get feedback early. Especailly since @dempseyatgithub had some ideas too.

#46

### CODE REVIEW

**Comment by mattmassicotte** [comment]
cc @hborla @ktoso

File: Guide.docc/Guidance.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Guidance
```
Comment (hborla) [inline Guide.docc/Guidance.md:1]: I think the title should be more specific. Most of the guide is guidance on specific topics. Maybe something like "Migration Strategy"?

**Review by hborla — COMMENTED** [review]
I think this is a great start. I think we should also mention some of the "upcoming features" that are very high value to enable _before_ enabling complete concurrency checking because they have the potential to eliminate a huge subset of the complete concurrency warnings. The two that come to mind are `GlobalConcurrency` and `InferSendableFromCaptures`.

File: Guide.docc/Guidance.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+**Don't panic.**
+
+Frequently, you'll find yourself making substantial progress with just a few
+changes.
+And as you do, your mental model of how the Swift concurrency system works
+will develop just as rapidly.
+
+## Strategy
+
+Of course, there is no one single approach that will work for all projects.
```
Comment (ktoso) [inline Guide.docc/Guidance.md:22]: I think we tend to avoid "of course" and similar like "it's obvious that..." sine for someone it may not be of course but be a surprise etc, so to not make them feel back avoid that wording?

```suggestion
There is no one single approach that will work for all projects.
```

File: Guide.docc/Guidance.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+But there is a general strategy that is often very effective.
```
Comment (ktoso) [inline Guide.docc/Guidance.md:23]: Isn't this a single sentence with the previous bit? `There is no..., but`
```suggestion
, but there is a general strategy that is often very effective.
```

File: Guide.docc/Guidance.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+- Select a module
+- Enable complete checking with Swift 5
+- Address warnings
+
+This process will be inherently _iterative_.
+Even a single change in one module can have a large impact on the state of the
+project as a whole.
+
+## Module Selection
```
Comment (ktoso) [inline Guide.docc/Guidance.md:34]:
```suggestion
## Migrating from the outside-in
```

```suggestion
## Start with leaf modules
```

Or something like that?

File: Guide.docc/Guidance.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+It is often easiest to start with the outer-most root module in a project.
+This, by definition, is not a depenency of any other module.
+Changes here can only have local effects, making it easier to
+keep your changes contained.
+
+## Enable Complete Checking
+
+Don't move directly from Swift 5 with no checking to the Swift 6 language mode.
```
Comment (ktoso) [inline Guide.docc/Guidance.md:43]: Maybe rephrase into 
```suggestion
You don't have to directly from Swift 5 with no checking to the Swift 6 language mode.
```

**Review by ktoso — COMMENTED** [review]
Good start, some minor wording ideas; pick or not, your call :)

**Review by ktoso — APPROVED** [review]

File: Guide.docc/Guidance.md (hunk 1 above)
Comment (mattmassicotte) [reply Guide.docc/Guidance.md:1]: ha yeah

File: Guide.docc/Guidance.md (hunk 2 above)
Comment (mattmassicotte) [reply Guide.docc/Guidance.md:22]: Yes, I actually really dislike this kind of wording too. Not sure how I let that happen.
